crates/store/src/filesystem/mod.rs

```diff
@@ -0,0 +1,412 @@
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+const KEY_FILE: &str = "key.bson";
+const META_FILE: &str = "meta.bson";
+const CONTENT_LINK_FILE: &str = "content.bin";
+
+/// Filesystem backed storage.
+///
+/// Layout:
+///
+///```text
+///root/
+///  storename/
+///    len-sha256/
+///      key/
+///        (key.bson length)-(key.bson checksum)/
+///          key.bson - unique
+///          meta.bson - possible duplicate
+///          content.bin - symlink to value/*.bin file
+///      value/
+///        (length)-(checksum).bin - unique
+///```
+#[derive(Clone)]
+pub struct Filesystem<K> {
+    root: Dir,
+    key_root: String,
+    value_root: String,
+    phantom: PhantomData<K>,
+}
+
+impl<K> Filesystem<K> {
+    async fn get_keys(&self) -> Vec<Result<K, Error>>
+    where
+        K: DeserializeOwned + 'static + Sync + Send + Clone,
+    {
+        let mut keys = vec![];
+
+        if let Ok(entries) = self.root.read_dir(&self.key_root).await {
+            for key_dir in entries {
+                if let Ok(Ok(key_dir)) = key_dir.map(|dir| dir.open_dir()) {
+                    if let Ok(entries) = key_dir.entries().await {
+                        for key_file in entries {
+                            let key_file = match key_file {
+                                Err(_) => continue,
+                                Ok(key_file) => key_file,
+                            };
+
+                            if utility::dir_entry_contains(&key_file, "key.bson") {
+                                let key_file = key_file.open();
+
+                                keys.push(async {
+                                    self.read_bincode_file(&mut key_file.map_err(Error::Io)?)
+                                        .await
+                                });
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        join_all(keys.into_iter()).await
+    }
+
+    async fn get_key_dir(&self, expected_identify: &str) -> Result<PathBuf, Error>
+    where
+        K: Serialize + Sync + Send + Eq + Hash,
+    {
+        let key_dir: PathBuf = format!("{}/{}", self.key_root, expected_identify).into();
+
+        if !self.root.exists(&key_dir).await {
+            return Err(Error::NotFound);
+        }
+
+        let mut key_file = key_dir.clone();
+        key_file.push(KEY_FILE);
+
+        let actual_identity = utility::len_sha256(&self.read_bytes(&key_file).await?);
+
+        if expected_identify != actual_identity {
+            return Err(Error::InvalidKeyFile);
+        }
+
+        Ok(key_dir)
+    }
+
+    async fn read_bytes<P: AsRef<Path>>(&self, path: P) -> Result<Vec<u8>, Error> {
+        if !self.root.exists(&path).await {
+            Err(Error::NotFound)
+        } else {
+            self.root.read(path).await.map_err(Error::Io)
+        }
+    }
+
+    async fn read_bytes_file(&self, file: &mut File) -> Result<Vec<u8>, Error> {
```

Comment:
Why have both `file` and non-`file` variants of these functions?

Reply:
Just to try and cut down on code repetition. Sometimes I have a file handle cheaply available and other times I only have the path.

Reply:
I'm confused at why `read_bytes_file` exists, since it doesn't actually make use of self at all. Would it be better as a method on File, so you can do `my_file.read_bytes()`?

Reply:
I moved it outside of Filesystem. I don't think my_file.read_bytes() exists in std so I don't think they would want me to add that.

Reply:
I removed read_bytes_file and read_bytes and replaced them with

```rust
async fn open<P: AsRef<Path>>(root: &Dir, path: P) -> Result<File, Error> {
    if !root.exists(&path).await {
        return Err(Error::NotFound);
    }

    root.open(path).map_err(Error::Io).await
}

async fn read<R: AsyncReadExt + Unpin>(reader: &mut R) -> Result<Vec<u8>, Error> {
    let mut buffer = vec![];
    reader.read_to_end(&mut buffer).await.map_err(Error::Io)?;
    Ok(buffer)
}
```

so they share an interface and are a bit simpler